packages/common/src/store/buy-crypto/sagas.ts

```diff
@@ -0,0 +1,497 @@
+import {
+  NATIVE_MINT,
+  createAssociatedTokenAccountIdempotentInstruction,
+  createCloseAccountInstruction,
+  createSyncNativeInstruction,
+  getAssociatedTokenAddressSync
+} from '@solana/spl-token'
+import {
+  LAMPORTS_PER_SOL,
+  PublicKey,
+  SystemProgram,
+  TransactionInstruction
+} from '@solana/web3.js'
+import { call, put, race, select, take, takeLatest } from 'typed-redux-saga'
+
+import { Name } from 'models/Analytics'
+import { ErrorLevel } from 'models/ErrorReporting'
+import { jupiterInstance, parseJupiterInstruction } from 'services/Jupiter'
+import {
+  IntKeys,
+  MEMO_PROGRAM_ID,
+  MintName,
+  createUserBankIfNeeded,
+  createVersionedTransaction,
+  getRootSolanaAccount,
+  getSolanaConnection,
+  pollForBalanceChange,
+  relayVersionedTransaction
+} from 'services/index'
+import {
+  onrampCanceled,
+  onrampFailed,
+  onrampSucceeded,
+  buyCryptoViaSol,
+  buyCryptoCanceled,
+  buyCryptoFailed,
+  buyCryptoSucceeded
+} from 'store/buy-crypto/slice'
+import { getBuyUSDCRemoteConfig } from 'store/buy-usdc'
+import { getContext } from 'store/commonStore'
+import { getFeePayer } from 'store/solana/selectors'
+import { TOKEN_LISTING_MAP } from 'store/ui/buy-audio/constants'
+import { setVisibility } from 'store/ui/modals/parentSlice'
+import { initializeStripeModal } from 'store/ui/stripe-modal/slice'
+
+import { BuyAudioStage, OnRampProvider } from '..'
+
+import { BuyCryptoConfig } from './types'
+
+export enum BuyCryptoErrorCode {
+  BAD_AMOUNT = 'BadAmount',
+  BAD_TOKEN = 'BadToken',
+  BAD_PROVIDER = 'BadProvider',
+  BAD_FEE_PAYER = 'BadFeePayer',
+  SWAP_ERROR = 'SwapError',
+  ON_RAMP_ERROR = 'OnRampError',
+  UNKNOWN = 'UnknownError'
+}
+
+class BuyCryptoError extends Error {
+  name = 'BuyCryptoError'
+  constructor(public code: BuyCryptoErrorCode, message: string) {
+    super(`${code}: ${message}`)
+  }
+}
```

Comment:
Feels like the error codes and error should be in the `types` module. But I guess that only really matters if you plan to use them outside this module.